### ISSUE TITLE
modules.yaml: "exclude_implicits" fails

### ISSUE THREAD

**Issue by G-Ragghianti** [LLNL/spack#34730]
### Steps to reproduce

Dockerfile:
```
FROM spack/ubuntu-focal
COPY modules.yaml /opt/spack/etc/spack/
RUN spack -d install xz
```

modules.yaml:
```
modules:
  default:
    enable:
      - tcl
    tcl:
      exclude_implicits: true
```

### Error message

```
==> [2022-12-30-13:48:51.482921] 	EXCLUDED_AS_IMPLICIT : xz@5.2.7%gcc@9.4.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-zen2/rflyw4i
```

The module files are not written even for explicitly installed specs.  However, doing "spack module tcl refresh" does write our the correct specs.  This occurs because the "exclude_implicits" config in modules.yaml activates code in lib/spack/spack/modules/common.py which calls "spec._installed_explicitly()".  This returns "None" which makes the code think that everything is implicitly installed. It returns "None" because `_installed_explicitly()` in spec.py fails to find the spec in "spack.store.db.get_record(self)" but then catches its own KeyError and returns "None". 

This works correctly when inside an environment.

### Information on your system

* **Spack:** 0.20.0.dev0
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-zen2
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by TomMelt** [comment]
I have the same issue. When I use `exclude_implicits: true` in my `module.yaml` file the auto-generation of modulefiles breaks.

```
==> [2023-01-17-18:35:38.074093]        EXCLUDED_AS_IMPLICIT : sphng@default%intel@2021.6.0+mpi arch=linux-ubuntu22.04-skylake/qitjtzh
==> [2023-01-17-18:35:38.074435]        NOT WRITING: sphng@default%intel@2021.6.0+mpi arch=linux-ubuntu22.04-skylake/qitjtzh [EXCLUDED]
==> [2023-01-17-18:35:38.074506] NO MODULE WRITTEN: list of enabled module files is empty
```

If I remove `exclude_implicits: true` from my `module.yaml` file then it works fine.

* **Spack:** 0.19.0.dev0 (1313bc99a6402d4f6ece5d0a9987b2ff7b1dfeee)
* **Python:** 3.10.6
* **Platform:** linux-ubuntu22.04-skylake
* **Concretizer:** clingo

**Comment by xdelaruelle** [comment]
I am trying to fix #32013 and I also face this issue.

From what I have been able to track down, when installing a software, it goes through `_install_task` in [lib/spack/spack/installer.py](https://github.com/spack/spack/blob/develop/lib/spack/spack/installer.py#L1264-L1277).

Module generation relies on the `explicit` field of database record for spec to know if software is explicit asked by user.

If we look at `_install_task`, module generation is done when calling to `spack.build_environment.start_build_process`. Database record for spec is set after this call, so module generation step do not have the information when it proceed.

**Comment by xdelaruelle** [comment]
#12105 describes the same issue
